File: yogi-cpp/yogi/timestamp.cpp
#include "timestamp.hpp"

#include <ctime>
#include <sstream>
#include <iomanip>
#include <mutex>


namespace yogi {

Timestamp::Timestamp(clock_type::time_point timePoint)
: m_timePoint(timePoint)
{
}

Timestamp Timestamp::now()
{
    return Timestamp(clock_type::now());
}

std::time_t Timestamp::to_time_t() const
{
    return clock_type::to_time_t(m_timePoint);
}

std::tm Timestamp::to_tm() const
{
    static std::mutex mutex;
    std::lock_guard<std::mutex> lock(mutex);

    auto time = to_time_t();
    auto tm = *std::gmtime(&time);

    return tm;
}

std::chrono::nanoseconds Timestamp::time_since_epoch() const
{
    return std::chrono::duration_cast<std::chrono::nanoseconds>(m_timePoint.time_since_epoch());
}

unsigned Timestamp::milliseconds() const
{
    return (time_since_epoch().count() / 1000000) % 1000;
}

unsigned Timestamp::microseconds() const
{
    return (time_since_epoch().count() / 1000) % 1000;
}

unsigned Timestamp::nanoseconds() const
{
    return time_since_epoch().count() % 1000;
}

std::string Timestamp::to_string() const
{
    return to_string(MILLISECONDS);
}

std::string Timestamp::to_string(Precision precision) const
{
    auto tm = to_tm();

    std::stringstream ss;
    char buffer[32];
    std::strftime(buffer, sizeof(buffer), "%d/%m/%Y %T", &tm);
    ss << buffer;

    if (precision >= MILLISECONDS) {
        ss << '.' << std::setw(3) << std::setfill('0') << milliseconds();
    }

    if (precision >= MICROSECONDS) {
        ss << '.' << std::setw(3) << std::setfill('0') << microseconds();
    }

    if (precision >= NANOSECONDS) {
        ss << '.' << std::setw(3) << std::setfill('0') << nanoseconds();
    }

    return ss.str();
}

} // namespace yogi

std::ostream& operator<< (std::ostream& os, const yogi::Timestamp& timestamp)
{
    return os << timestamp.to_string();
}

File: yogi-python/yogi/private/timestamp.py
from enum import Enum
import time


class Timestamp:
    class Precision(Enum):
        SECONDS = 0
        MILLISECONDS = 1
        MICROSECONDS = 2
        NANOSECONDS = 3

    @classmethod
    def now(cls):
        timestamp = Timestamp()
        timestamp._time = time.time()
        timestamp._ns_since_epoch = int(timestamp._time * 1e9)
        return timestamp

    @property
    def ns_since_epoch(self) -> int:
        return self._ns_since_epoch

    @property
    def milliseconds(self) -> int:
        return int(self._ns_since_epoch / 1000000) % 1000

    @property
    def microseconds(self) -> int:
        return int(self._ns_since_epoch / 1000) % 1000

    @property
    def nanoseconds(self) -> int:
        return self._ns_since_epoch % 1000

    def to_string(self, precision: Precision = Precision.MILLISECONDS) -> str:
        s = time.strftime('%d/%m/%Y %T', time.localtime(int(self._time)))

        if precision.value >= self.Precision.MILLISECONDS.value:
            s += '.{:03}'.format(self.milliseconds)
        if precision.value >= self.Precision.MICROSECONDS.value:
            s += '.{:03}'.format(self.microseconds)
        if precision.value >= self.Precision.NANOSECONDS.value:
            s += '.{:03}'.format(self.nanoseconds)

        return s

    def __str__(self):
        return self.to_string()
